// @ng-dynamic-forms/core/esm2020/lib/model/dynamic-input-control.model.mjs
import { __decorate } from "tslib";
import { DynamicFormValueControlModel } from "./dynamic-form-value-control.model";
import { serializable } from "../decorator/serializable.decorator";
import { isBoolean, isNumber } from "../utils/core.utils";
export class DynamicInputControlModel extends DynamicFormValueControlModel {
    constructor(config, layout) {
        super(config, layout);
        this.autoComplete = config.autoComplete ?? "on";
        this.autoFocus = isBoolean(config.autoFocus) ? config.autoFocus : false;
        this.maxLength = isNumber(config.maxLength) ? config.maxLength : null;
        this.minLength = isNumber(config.minLength) ? config.minLength : null;
        this.placeholder = config.placeholder ?? "";
        this.prefix = config.prefix ?? null;
        this.readOnly = isBoolean(config.readOnly) ? config.readOnly : false;
        this.spellCheck = isBoolean(config.spellCheck) ? config.spellCheck : false;
        this.suffix = config.suffix ?? null;
    }
}
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "autoComplete", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "autoFocus", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "maxLength", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "minLength", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "placeholder", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "prefix", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "readOnly", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "spellCheck", void 0);
__decorate([
    serializable()
], DynamicInputControlModel.prototype, "suffix", void 0);
