// @ng-dynamic-forms/core/esm2020/lib/model/dynamic-form-control.model.mjs
import { __decorate } from "tslib";
import { BehaviorSubject } from "rxjs";
import { serializable, serialize } from "../decorator/serializable.decorator";
import { isBoolean, isObject, isString } from "../utils/core.utils";
export class DynamicFormControlModel {
    constructor(config, layout = null) {
        this.parent = null;
        this.asyncValidators = config.asyncValidators ?? null;
        this.errorMessages = config.errorMessages ?? null;
        this.hidden = isBoolean(config.hidden) ? config.hidden : false;
        this.id = config.id;
        this.label = config.label ?? null;
        this.labelTooltip = config.labelTooltip ?? null;
        this.controlTooltip = config.controlTooltip ?? null;
        this.layout = layout;
        this.name = config.name ?? config.id;
        this.relations = Array.isArray(config.relations) ? config.relations : [];
        this.updateOn = isString(config.updateOn) ? config.updateOn : null;
        this.validators = config.validators ?? null;
        this._disabled = isBoolean(config.disabled) ? config.disabled : false;
        this.disabled$ = new BehaviorSubject(this._disabled);
        this.disabled$.subscribe(disabled => this._disabled = disabled);
        this.disabledChanges = this.disabled$.asObservable();
    }
    get disabled() {
        return this.disabled$.getValue();
    }
    set disabled(disabled) {
        this.disabled$.next(disabled);
    }
    get hasErrorMessages() {
        return isObject(this.errorMessages);
    }
    toJSON() {
        return serialize(this);
    }
}
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "asyncValidators", void 0);
__decorate([
    serializable("disabled")
], DynamicFormControlModel.prototype, "_disabled", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "errorMessages", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "hidden", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "id", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "label", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "labelTooltip", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "controlTooltip", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "layout", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "name", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "relations", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "updateOn", void 0);
__decorate([
    serializable()
], DynamicFormControlModel.prototype, "validators", void 0);
