// @ng-dynamic-forms/core/esm2020/lib/model/textarea/dynamic-textarea.model.mjs
import { __decorate } from "tslib";
import { DynamicInputControlModel } from "../dynamic-input-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isNumber } from "../../utils/core.utils";
export const DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA = "TEXTAREA";
export const DYNAMIC_FORM_TEXTAREA_WRAP_HARD = "hard";
export const DYNAMIC_FORM_TEXTAREA_WRAP_SOFT = "soft";
export class DynamicTextAreaModel extends DynamicInputControlModel {
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA;
        this.cols = isNumber(config.cols) ? config.cols : 20;
        this.rows = isNumber(config.rows) ? config.rows : 2;
        this.wrap = config.wrap ?? DYNAMIC_FORM_TEXTAREA_WRAP_SOFT;
    }
}
__decorate([
    serializable()
], DynamicTextAreaModel.prototype, "cols", void 0);
__decorate([
    serializable()
], DynamicTextAreaModel.prototype, "rows", void 0);
__decorate([
    serializable()
], DynamicTextAreaModel.prototype, "wrap", void 0);
__decorate([
    serializable()
], DynamicTextAreaModel.prototype, "type", void 0);
