// @ng-dynamic-forms/core/esm2020/lib/component/dynamic-form-control-with-template.component.mjs
import { Directive } from "@angular/core";
import { DynamicFormControlComponent } from "./dynamic-form-control.component";
import { isString } from "../utils/core.utils";
import * as i0 from "@angular/core";
import * as i1 from "../service/dynamic-form-layout.service";
import * as i2 from "../service/dynamic-form-validation.service";
// tslint:disable-next-line:directive-class-suffix
export class DynamicFormControlWithTemplateComponent extends DynamicFormControlComponent {
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
    }
    ngAfterViewInit() {
        this.layoutService
            .filterTemplatesByModel(this.model, this.templates)
            .forEach(template => this.bindTemplate(template));
    }
    bindTemplate(template) {
        if (isString(template.as) && this.templateDirectives.has(template.as)) {
            const property = this.templateDirectives.get(template.as);
            this.viewChild[property] = this.mapTemplate(template);
        }
    }
}
DynamicFormControlWithTemplateComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: DynamicFormControlWithTemplateComponent, deps: [{ token: i1.DynamicFormLayoutService }, { token: i2.DynamicFormValidationService }], target: i0.ɵɵFactoryTarget.Directive });
DynamicFormControlWithTemplateComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.2", type: DynamicFormControlWithTemplateComponent, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: DynamicFormControlWithTemplateComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.DynamicFormLayoutService }, { type: i2.DynamicFormValidationService }]; } });
