// mirador-share-plugin/lib/MiradorShareDialog.js
'use strict';

exports.__esModule = true;
exports.MiradorShareDialog = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Dialog = require('@material-ui/core/Dialog');

var _Dialog2 = _interopRequireDefault(_Dialog);

var _DialogActions = require('@material-ui/core/DialogActions');

var _DialogActions2 = _interopRequireDefault(_DialogActions);

var _DialogTitle = require('@material-ui/core/DialogTitle');

var _DialogTitle2 = _interopRequireDefault(_DialogTitle);

var _Divider = require('@material-ui/core/Divider');

var _Divider2 = _interopRequireDefault(_Divider);

var _Link = require('@material-ui/core/Link');

var _Link2 = _interopRequireDefault(_Link);

var _TextField = require('@material-ui/core/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _Typography = require('@material-ui/core/Typography');

var _Typography2 = _interopRequireDefault(_Typography);

var _Grid = require('@material-ui/core/Grid');

var _Grid2 = _interopRequireDefault(_Grid);

var _reactCopyToClipboard = require('react-copy-to-clipboard');

var _notistack = require('notistack');

var _manifests = require('mirador/dist/es/src/state/selectors/manifests');

var _config = require('mirador/dist/es/src/state/selectors/config');

var _ScrollIndicatedDialogContent = require('mirador/dist/es/src/containers/ScrollIndicatedDialogContent');

var _ScrollIndicatedDialogContent2 = _interopRequireDefault(_ScrollIndicatedDialogContent);

var _MiradorShareEmbed = require('./MiradorShareEmbed');

var _MiradorShareEmbed2 = _interopRequireDefault(_MiradorShareEmbed);

var _IiifIcon = require('./IiifIcon');

var _IiifIcon2 = _interopRequireDefault(_IiifIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var mapDispatchToProps = function mapDispatchToProps(dispatch, _ref) {
  var windowId = _ref.windowId;
  return {
    closeShareDialog: function closeShareDialog() {
      return dispatch({ type: 'CLOSE_WINDOW_DIALOG', windowId: windowId });
    }
  };
};

var mapStateToProps = function mapStateToProps(state, _ref2) {
  var windowId = _ref2.windowId;

  var miradorSharePluginConfig = state.config.miradorSharePlugin || {};
  var embedOption = miradorSharePluginConfig.embedOption || {};

  return {
    containerId: (0, _config.getContainerId)(state),
    displayEmbedOption: embedOption.enabled,
    displayShareLink: miradorSharePluginConfig.shareLink && miradorSharePluginConfig.shareLink.enabled,
    embedUrlReplacePattern: embedOption.embedUrlReplacePattern,
    embedIframeAttributes: embedOption.embedIframeAttributes,
    embedIframeTitle: embedOption.embedIframeTitle,
    manifestIdReplacePattern: miradorSharePluginConfig.shareLink && miradorSharePluginConfig.shareLink.manifestIdReplacePattern,
    iiifInfoLink: miradorSharePluginConfig.iiifInfoLink,
    manifestId: ((0, _manifests.getManifestoInstance)(state, { windowId: windowId }) || {}).id,
    open: state.windowDialogs[windowId] && state.windowDialogs[windowId].openDialog === 'share',
    syncIframeDimensions: embedOption.syncIframeDimensions
  };
};

/**
 * MiradorShareDialog ~
*/

var MiradorShareDialog = exports.MiradorShareDialog = function (_Component) {
  _inherits(MiradorShareDialog, _Component);

  function MiradorShareDialog(props) {
    _classCallCheck(this, MiradorShareDialog);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      shareLinkText: _this.shareLink()
    };

    _this.handleShareLinkChange = _this.handleShareLinkChange.bind(_this);
    return _this;
  }

  MiradorShareDialog.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    var manifestId = this.props.manifestId;

    if (manifestId !== prevProps.manifestId) {
      this.handleShareLinkChange(this.shareLink());
    }
  };

  MiradorShareDialog.prototype.dragAndDropUrl = function dragAndDropUrl() {
    var manifestId = this.props.manifestId;

    var baseUrl = manifestId;

    return baseUrl + '?manifest=' + manifestId;
  };

  MiradorShareDialog.prototype.whatIsThisLink = function whatIsThisLink() {
    var iiifInfoLink = this.props.iiifInfoLink;


    if (!iiifInfoLink) return null;

    return _react2.default.createElement(
      _react2.default.Fragment,
      null,
      ' ',
      _react2.default.createElement(
        _Link2.default,
        { href: iiifInfoLink },
        'What is IIIF?'
      )
    );
  };

  MiradorShareDialog.prototype.shareLink = function shareLink() {
    var _props = this.props,
        manifestId = _props.manifestId,
        manifestIdReplacePattern = _props.manifestIdReplacePattern;

    if (!manifestId) return null;

    return manifestId.replace(manifestIdReplacePattern[0], manifestIdReplacePattern[1]);
  };

  MiradorShareDialog.prototype.handleShareLinkChange = function handleShareLinkChange(value) {
    this.setState({
      shareLinkText: value
    });
  };

  /**
   * Returns the rendered component
  */


  MiradorShareDialog.prototype.render = function render() {
    var _this2 = this;

    var _props2 = this.props,
        classes = _props2.classes,
        closeShareDialog = _props2.closeShareDialog,
        containerId = _props2.containerId,
        displayEmbedOption = _props2.displayEmbedOption,
        displayShareLink = _props2.displayShareLink,
        embedIframeAttributes = _props2.embedIframeAttributes,
        embedIframeTitle = _props2.embedIframeTitle,
        embedUrlReplacePattern = _props2.embedUrlReplacePattern,
        manifestId = _props2.manifestId,
        open = _props2.open,
        syncIframeDimensions = _props2.syncIframeDimensions;
    var shareLinkText = this.state.shareLinkText;


    if (!open) return _react2.default.createElement(_react2.default.Fragment, null);

    return _react2.default.createElement(
      _Dialog2.default,
      {
        container: document.querySelector('#' + containerId + ' .mirador-viewer'),
        onClose: closeShareDialog,
        open: open
      },
      _react2.default.createElement(
        _DialogTitle2.default,
        { disableTypography: true, className: classes.h2 },
        _react2.default.createElement(
          _Typography2.default,
          { variant: 'h2' },
          'Share'
        )
      ),
      _react2.default.createElement(_notistack.SnackbarProvider, {
        maxSnack: 1,
        anchorOrigin: {
          vertical: 'top',
          horizontal: 'center'
        }
      }),
      _react2.default.createElement(
        _ScrollIndicatedDialogContent2.default,
        null,
        displayShareLink && _react2.default.createElement(
          _react2.default.Fragment,
          null,
          _react2.default.createElement(
            _Typography2.default,
            { className: classes.h3, variant: 'h3' },
            'Share link'
          ),
          _react2.default.createElement(
            'div',
            { className: classes.inputContainer },
            _react2.default.createElement(_TextField2.default, {
              defaultValue: shareLinkText,
              fullWidth: true,
              variant: 'filled',
              onChange: function onChange(e) {
                return _this2.handleShareLinkChange(e && e.target && e.target.value);
              },
              inputProps: { 'aria-label': 'Share link URL', className: classes.shareLinkInput }
            }),
            ' ',
            _react2.default.createElement(
              _reactCopyToClipboard.CopyToClipboard,
              { text: shareLinkText },
              _react2.default.createElement(
                _Button2.default,
                {
                  className: classes.copyButton,
                  variant: 'outlined',
                  color: 'primary',
                  'aria-label': 'Copy link to clipboard',
                  onClick: function onClick() {
                    return (0, _notistack.enqueueSnackbar)(_react2.default.createElement(
                      _Typography2.default,
                      { variant: 'body1' },
                      'Link copied to clipboard!'
                    ), { variant: 'success' });
                  }
                },
                'Copy'
              )
            )
          ),
          _react2.default.createElement(_Divider2.default, null)
        ),
        displayEmbedOption && _react2.default.createElement(
          _react2.default.Fragment,
          null,
          _react2.default.createElement(
            _Typography2.default,
            { className: classes.h3, variant: 'h3' },
            'Embed'
          ),
          _react2.default.createElement(_MiradorShareEmbed2.default, {
            embedIframeAttributes: embedIframeAttributes,
            embedIframeTitle: embedIframeTitle,
            embedUrlReplacePattern: embedUrlReplacePattern,
            syncIframeDimensions: syncIframeDimensions,
            manifestId: manifestId
          }),
          _react2.default.createElement(_Divider2.default, null)
        ),
        _react2.default.createElement(
          _Typography2.default,
          { className: classes.h3, variant: 'h3' },
          'Add to another viewer'
        ),
        _react2.default.createElement(
          _Grid2.default,
          { container: true, spacing: 1, className: classes.grid },
          _react2.default.createElement(
            _Grid2.default,
            { item: true, xs: true },
            _react2.default.createElement(
              _Typography2.default,
              { variant: 'body1' },
              'Drag & drop IIIF icon to add this resource to any IIIF viewer.'
            ),
            _react2.default.createElement(
              _Link2.default,
              { href: this.dragAndDropUrl(), className: classes.iiifLink },
              _react2.default.createElement(_IiifIcon2.default, { className: classes.iiifIcon })
            )
          ),
          _react2.default.createElement(
            _Grid2.default,
            { item: true, xs: 1 },
            _react2.default.createElement(
              _Typography2.default,
              { variant: 'body1' },
              'or'
            )
          ),
          _react2.default.createElement(
            _Grid2.default,
            { item: true, xs: true },
            _react2.default.createElement(
              _Typography2.default,
              { variant: 'body1' },
              'Copy & paste the resource\'s manifest into any IIIF viewer.'
            ),
            _react2.default.createElement(
              _reactCopyToClipboard.CopyToClipboard,
              { text: this.dragAndDropUrl() },
              _react2.default.createElement(
                _Button2.default,
                {
                  className: classes.copyButton,
                  variant: 'outlined',
                  color: 'primary',
                  'aria-label': 'Copy manifest to clipboard',
                  onClick: function onClick() {
                    return (0, _notistack.enqueueSnackbar)(_react2.default.createElement(
                      _Typography2.default,
                      { variant: 'body1' },
                      'Manifest copied to clipboard!'
                    ), { variant: 'success' });
                  }
                },
                'Copy'
              )
            )
          )
        ),
        _react2.default.createElement(
          _Typography2.default,
          { variant: 'body1' },
          this.whatIsThisLink()
        )
      ),
      _react2.default.createElement(
        _DialogActions2.default,
        null,
        _react2.default.createElement(
          _Button2.default,
          { onClick: closeShareDialog, color: 'primary' },
          'Close'
        )
      )
    );
  };

  return MiradorShareDialog;
}(_react.Component);

MiradorShareDialog.propTypes = process.env.NODE_ENV !== "production" ? {
  classes: _propTypes2.default.shape({
    copyButton: _propTypes2.default.string,
    h2: _propTypes2.default.string,
    h3: _propTypes2.default.string,
    iiifIcon: _propTypes2.default.string,
    iiifLink: _propTypes2.default.string,
    inputContainer: _propTypes2.default.string,
    shareLinkInput: _propTypes2.default.string,
    grid: _propTypes2.default.string
  }).isRequired,
  closeShareDialog: _propTypes2.default.func.isRequired,
  containerId: _propTypes2.default.string.isRequired,
  displayEmbedOption: _propTypes2.default.bool,
  displayShareLink: _propTypes2.default.bool,
  iiifInfoLink: _propTypes2.default.string,
  embedIframeAttributes: _propTypes2.default.string,
  embedIframeTitle: _propTypes2.default.string,
  embedUrlReplacePattern: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(RegExp)])),
  manifestIdReplacePattern: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(RegExp)])),
  manifestId: _propTypes2.default.string,
  open: _propTypes2.default.bool,
  syncIframeDimensions: _propTypes2.default.shape({
    height: _propTypes2.default.shape({ param: _propTypes2.default.string }),
    width: _propTypes2.default.shape({ param: _propTypes2.default.string })
  })
} : {};

MiradorShareDialog.defaultProps = {
  displayEmbedOption: false,
  displayShareLink: false,
  iiifInfoLink: 'https://iiif.io',
  embedIframeAttributes: 'allowfullscreen frameborder="0"',
  embedIframeTitle: 'Image viewer',
  embedUrlReplacePattern: [],
  manifestId: '',
  manifestIdReplacePattern: [],
  open: false,
  syncIframeDimensions: {}
};

var styles = function styles(theme) {
  return {
    copyButton: {
      marginLeft: theme.spacing()
    },
    h2: {
      paddingBottom: 0
    },
    h3: {
      marginTop: '20px'
    },
    iiifLink: {
      marginRight: '10px'
    },
    iiifIcon: {
      verticalAlign: 'text-bottom',
      cursor: 'grab',
      paddingTop: '12px'
    },
    inputContainer: {
      alignItems: 'flex-end',
      display: 'flex',
      flexDirection: 'row',
      marginBottom: theme.spacing()
    },
    shareLinkInput: {
      paddingTop: '12px'
    },
    grid: {
      textAlign: 'center',
      paddingTop: '12px'
    }
  };
};

exports.default = {
  target: 'Window',
  mode: 'add',
  name: 'MiradorShareDialog',
  component: (0, _styles.withStyles)(styles)(MiradorShareDialog),
  mapDispatchToProps: mapDispatchToProps,
  mapStateToProps: mapStateToProps
};
