// @ng-dynamic-forms/core/esm2020/lib/model/select/dynamic-select.model.mjs
import { __decorate } from "tslib";
import { DynamicOptionControlModel } from "../dynamic-option-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isFunction } from "../../utils/core.utils";
export const DYNAMIC_FORM_CONTROL_TYPE_SELECT = "SELECT";
export class DynamicSelectModel extends DynamicOptionControlModel {
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SELECT;
        this.compareWithFn = isFunction(config.compareWithFn) ? config.compareWithFn : Object.is;
        this.filterable = isBoolean(config.filterable) ? config.filterable : false;
        this.multiple = isBoolean(config.multiple) ? config.multiple : false;
        this.placeholder = config.placeholder ?? "";
        this.prefix = config.prefix ?? null;
        this.suffix = config.suffix ?? null;
    }
    select(...indices) {
        this.value = this.multiple ? indices.map(index => this.get(index).value) : this.get(indices[0]).value;
    }
}
__decorate([
    serializable()
], DynamicSelectModel.prototype, "filterable", void 0);
__decorate([
    serializable()
], DynamicSelectModel.prototype, "multiple", void 0);
__decorate([
    serializable()
], DynamicSelectModel.prototype, "placeholder", void 0);
__decorate([
    serializable()
], DynamicSelectModel.prototype, "prefix", void 0);
__decorate([
    serializable()
], DynamicSelectModel.prototype, "suffix", void 0);
__decorate([
    serializable()
], DynamicSelectModel.prototype, "type", void 0);
